runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,435 @@
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+      };
+    }
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public DoFn.InputProvider<InputT> inputProvider() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public DoFn.OutputReceiver<OutputT> outputReceiver() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public TrackerT restrictionTracker() {
+          return tracker;
+        }
+      };
+    }
+  }
+
+  /** Splits the restriction using the given {@link DoFn.SplitRestriction} method. */
+  private static class SplitRestrictionFn<InputT, RestrictionT>
+      extends DoFn<
+          ElementRestriction<InputT, RestrictionT>, ElementRestriction<InputT, RestrictionT>> {
+    private final DoFn<InputT, ?> fn;
+    private transient DoFnInvoker<InputT, ?> invoker;
+
+    SplitRestrictionFn(DoFn<InputT, ?> fn) {
+      this.fn = fn;
+    }
+
+    @Setup
+    public void setup() {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      List<RestrictionT> parts =
```

Comment:
Should we worry about the size of this list? If so, would it make sense to call wire up the invocation of the split restriction to immediately call output?